modelscope_agent/agent.py

```diff
@@ -1,377 +1,182 @@
         """
-        replace action_args in str to Image/Video/Audio Wrapper, so that tool can handle them
+        special_func_token = '\nAction:'
+        special_args_token = '\nAction Input:'
+        special_obs_token = '\nObservation:'
+        func_name, func_args = None, None
+        i = text.rfind(special_func_token)
+        j = text.rfind(special_args_token)
+        k = text.rfind(special_obs_token)
+        if 0 <= i < j:  # If the text has `Action` and `Action input`,
+            if k < j:  # but does not contain `Observation`,
+                # then it is likely that `Observation` is ommited by the LLM,
+                # because the output text may have discarded the stop word.
+                text = text.rstrip() + special_obs_token  # Add it back.
+            k = text.rfind(special_obs_token)
+            func_name = text[i + len(special_func_token):j].strip()
+            func_args = text[j + len(special_args_token):k].strip()
+            text = text[:k]  # Discard '\nObservation:'.
+
+        return (func_name is not None), func_name, func_args, text
+
+    def _detect_tool_by_func_call(self, message: Dict):
         """
-        parsed_action_args = {}
-        for name, arg in action_args.items():
-            try:
-                true_arg = self.agent_state.get(arg, arg)
-            except Exception as e:
-                logger.error(
-                    error=
-                    f'Error when parsing action args: {e}, using fall back')
-                true_arg = arg
-            parsed_action_args[name] = true_arg
-        return parsed_action_args
+        A built-in tool call detection for func_call format
 
-    def parse_exec_result(self, exec_result, *args, **kwargs):
-        """
-        update exec result to agent state.
-        key is the str representation of the result.
         """
-        for k, v in exec_result.items():
-            self.agent_state[str(v)] = v
+        func_name = None
+        func_args = None
+        if 'function_call' in message and message['function_call']:
+            func_call = message['function_call']
+            func_name = func_call.get('name', '')
+            func_args = func_call.get('arguments', '')
+        text = message.get('content', '')
+
+        return (func_name is not None), func_name, func_args, text
+
+    def send(self,
+             message: Union[Dict, str],
+             recipient: 'Agent',
+             request_reply: Optional[bool] = None,
+             **kwargs):
+        recipient.receive(message, self, request_reply)
+
+    def receive(self,
+                message: Union[Dict, str],
+                sender: 'Agent',
+                request_reply: Optional[bool] = None,
+                **kwargs):
+        if request_reply is False or request_reply is None:
+            return
+        reply = self.run(message, sender=sender, **kwargs)
+        if reply is not None:
+            self.send(reply, sender, **kwargs)
```

Comment:
these two interface are not used and referenced, may be added later, delete them now

Reply:
Deleted